### PR TITLE
r.viewshed.exposure: New addon to compute visual exposure

### PR DESCRIPTION
**Pull request by zofie-cimburova**

### CODE REVIEW

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.html [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+  The visual exposure in forest and rural landscapes:
+  An algorithm and a GIS tool.
+  <i>Landscape and Urban Planning</i> 101, 52-58.
+  <a href="https://doi.org/10.1016/j.landurbplan.2010.11.018">
+  https://doi.org/10.1016/j.landurbplan.2010.11.018</a></li>
+<li>Fisher, P., 1994.
+  Probable and fuzzy models of the viewshed operation,
+  in: Worboys, M.F. (Ed.), <i>Innovations in GIS</i>.
+  Taylor & Francis, London, pp. 161-176.</li>
+<li>Grêt-Regamey, A., Bishop, I.D., Bebi, P., 2007.
```
Comment (neteler) [inline src/raster/r.viewshed.exposure/r.viewshed.exposure.html:197]:
```suggestion
<li>Grêt-Regamey, A., Bishop, I.D., Bebi, P., 2007.
```

Please use HTML code for ê etc.

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.html [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+<i>visual magnitude function</i> and <i>solid angle function</i>.
+<p>
+
+In <i>distance decay function</i>, the contribution of an exposure source pixel
+<i>xi</i> to visual exposure at the observer pixel decreases in proportion to
+the square of distance between the exposure source pixel and the observer:
+<i>D(xi) = A/v<sup>2</sup></i>; <i>A</i> is the area of the exposure source
+pixel, <i>v</i> is the distance between the exposure source pixel and
+the observer.
+See Grêt-Regamey et al. (2007) and Chamberlain and Meitner (2013) for more
```
Comment (ninsbl) [inline src/raster/r.viewshed.exposure/r.viewshed.exposure.html:65]:
```suggestion
See Gr&ecirc;t-Regamey et al. (2007) and Chamberlain and Meitner (2013) for more
```

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.html (hunk 1 above)
Comment (ninsbl) [inline src/raster/r.viewshed.exposure/r.viewshed.exposure.html:197]:
```suggestion
<li>Gr&ecirc;t-Regamey, A., Bishop, I.D., Bebi, P., 2007.
```

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.html [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+  Viewshed-based modelling of visual exposure to urban greenery -
+  an efficient GIS tool for practical appliactions (in press).
+  Landscape and Urban Planning.
+</li>
+<li>Chamberlain, B.C., Meitner, M.J., 2013.
+  A route-based visibility analysis for landscape management.
+  <i>Landscape and Urban Planning</i> 111, 13-24.
+  <a href="https://doi.org/10.1016/j.landurbplan.2012.12.004">
+    https://doi.org/10.1016/j.landurbplan.2012.12.004 </a> </li>
+<li>Domingo-Santos, J.M., de Villarán, R.F., Rapp-Arrarás, Í., de Provens,
```
Comment (ninsbl) [inline src/raster/r.viewshed.exposure/r.viewshed.exposure.html:186]:
```suggestion
<li>Domingo-Santos, J.M., de Villar&aacute;n, R.F., Rapp-Arrar&aacute;s, &Iacute;., de Provens,
```

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.html [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+<div align="center" style="margin: 10px">
+<a href="r_viewshed_exposure_workflow.png">
+<img src="r_viewshed_exposure_workflow.png" width="489" height="545" alt="r.viewshed.exposure workflow" border="0">
+</a><br>
+<i>Processing workflow</i>
+</div>
+
+<h4>1. Random sampling of exposure source raster map with vector points</h4>
+To improve computational efficiency, the exposure source raster map is
+randomly sampled with defined density (0-100%; option <b>sample_density</b>).
```
Comment (ninsbl) [inline src/raster/r.viewshed.exposure/r.viewshed.exposure.html:30]:
```suggestion
randomly sampled with defined density (0-100&#37;; option <b>sample_density</b>).
```

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.html [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+viewshed computation. Option <b>cores</b> specifies the number of cores used in
+parallel processing. In parallel processing, the computation of individual
+viewsheds is randomly distributed across the specified cores.
+
+
+<h2>EXAMPLES</h2>
+Computation of visual exposure to major roads in South-West Wake county,
+North Carolina. Input data are a terrain model and a raster map of major roads
+from NC dataset. Viewshed parametrisation function is set to none (example 1)
+and solid angle (example 2). Sampling density is set to 50%,
```
Comment (ninsbl) [inline src/raster/r.viewshed.exposure/r.viewshed.exposure.html:113]:
```suggestion
and solid angle (example 2). Sampling density is set to 50&#37;,
```

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.html [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+
+In <i>fuzzy viewshed function</i>, the contribution of an exposure source pixel
+<i>xi</i> to visual exposure at the observer pixel depends on the distance
+between the exposure source pixel and the observer and the radius of perfect
+clarity.
+See Fisher (1994) and Ogburn (2006) for more details.
+<p>
+
+In <i>visual magnitude function</i>, the contribution of an exposure source pixel
+<i>xi</i> to visual exposure at the observer pixel depends on the pixel’s slope,
```
Comment (ninsbl) [inline src/raster/r.viewshed.exposure/r.viewshed.exposure.html:77]:
```suggestion
<i>xi</i> to visual exposure at the observer pixel depends on the pixel&#39;s slope,
```

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.html [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+ aspect and distance relative to the observer.
+See Chamberlain and Meitner (2013) for more details.
+<p>
+
+In <i>solid angle function</i>, the contribution of an exposure source pixel
+<i>xi</i> to visual exposure at the observer pixel is calculated as a solid
+angle, i.e. the area (in sterradians) of the observer's eye retina covered by
```
Comment (ninsbl) [inline src/raster/r.viewshed.exposure/r.viewshed.exposure.html:84]:
```suggestion
angle, i.e. the area (in sterradians) of the observer&#39;s eye retina covered by
```

**Review by ninsbl — COMMENTED** [review]
Suggestions for HTML codes

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.html (hunk 1 above)
Comment (ninsbl) [reply src/raster/r.viewshed.exposure/r.viewshed.exposure.html]: Thanks. Should be solved now.

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.py [hunk 8]
```diff
@@ -0,0 +1,1409 @@
...
+    }
+
+    # Split target points to chunks for each core
+    target_pnts = np.array_split(target_pts_np, cores)
+
+    # Combine each chunk with dictionary
+    combo = list(zip(itertools.repeat(global_vars), target_pnts))
+
+    # Calculate partial cummulative viewshed
+    pool = Pool(cores)
```
Comment (petrasovaa) [inline src/raster/r.viewshed.exposure/r.viewshed.exposure.py:1370]: Consider using context manager

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.py [hunk 9]
```diff
@@ -0,0 +1,1409 @@
...
+            target_elevation=v_elevation,
+            max_distance=exp_range,
+            refraction_coeff=refr_coeff,
+            memory=int(round(memory / cores)),
+            quiet=True,
+            overwrite=True,
+            env=c_env,
+        )
+        vs.communicate()
+        clean_temp(vs.pid)
```
Comment (petrasovaa) [inline src/raster/r.viewshed.exposure/r.viewshed.exposure.py:389]: Maybe comment on this, is this some workaround for a bug, the temp files should be cleaned automatically, no?

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.py [hunk 10]
```diff
@@ -0,0 +1,1409 @@
...
+
+        np_viewshed = raster2numpy(tmp_vs).astype(np.single)
+        np_viewshed[np_viewshed == -2147483648] = np.nan
```
Comment (petrasovaa) [inline src/raster/r.viewshed.exposure/r.viewshed.exposure.py:392]: Could you comment on this?

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.py [hunk 11]
```diff
@@ -0,0 +1,1409 @@
...
+# % description: Computes visual exposure to defined exposure source using weighted parametrised cummulative viewshed analysis.
+# % keyword: raster
+# % keyword: viewshed
+# % keyword: line of sight
+# % keyword: LOS
+# % keyword: visual exposure
+# %end
+
+# %option G_OPT_R_INPUT
+# % key: dsm
```
Comment (petrasovaa) [inline src/raster/r.viewshed.exposure/r.viewshed.exposure.py:34]: Consider using G_OPT_R_ELEV (key would be then elevation) or at least change key to 'input' to match r.viewshed.

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.py [hunk 12]
```diff
@@ -0,0 +1,1409 @@
...
+        n_vect_ew_unit[0] * v_vect_ew_unit[:][0]
+        + n_vect_ew_unit[1] * v_vect_ew_unit[:][1]
+    )
+    theta = np.where(theta > math.pi / 2, theta - math.pi / 2, theta)
+
+    # 8. visual magnitude adjusted for distance weight
+    visual_magnitude = (
+        np.cos(beta) * np.cos(theta) * ((reg.nsres * reg.ewres) / (v_scal ** 2))
+    )
+    sys.stderr.flush()
```
Comment (petrasovaa) [inline src/raster/r.viewshed.exposure/r.viewshed.exposure.py:874]: Why is this needed here?

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.py [hunk 13]
```diff
@@ -0,0 +1,1409 @@
...
+    :type min_d: float
+    :param v_sample: Name of output vector map of sampling points
+    :type v_sample: string
+    :param seed: Random seed
+    :param seed: int
+    :return: Name of output vector map of sampling points
+    :rtype: string
+    """
+    # mask categories of raster map to be sampled from
+    grass.run_command(
```
Comment (petrasovaa) [inline src/raster/r.viewshed.exposure/r.viewshed.exposure.py:898]: I recall we were able to remove the need for using mask in r.sample.category, so I wonder if this would be applicable here as well?

**Review by petrasovaa — CHANGES_REQUESTED** [review]
The code looks great, I added mostly minor comments. I didn't actually run the code yet, will try to do that later, so I may still have more comments.
What is little frustrating is how unreadable the numpy portions are (at least to me), but there is probably nothing to do about that.

**Comment by ninsbl** [comment]
Thanks for the thorough review so far. Very good and valid comments. Looking forward to hear more after you ran the code...

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.py [hunk 14]
```diff
@@ -0,0 +1,1409 @@
...
+# % type: integer
+# % required: no
+# % key_desc: value
+# % label: Amount of memory to use in MB
+# % options: 1-
+# % description: 1-
+# % answer: 500
+# %end
+
+# %option
```
[the remark this answers is not in the document]
Comment (ninsbl) [reply src/raster/r.viewshed.exposure/r.viewshed.exposure.py]: Addressed in ef04cc8

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.py [hunk 15]
```diff
@@ -0,0 +1,1409 @@
...
+            grass.run_command(
+                "g.remove",
+                quiet=True,
+                type="raster",
+                name="MASK",
+                stderr=subprocess.DEVNULL,
+                flags="f",
+            )
+        except Exception:
+            pass
```
[the remark this answers is not in the document]
Comment (ninsbl) [reply src/raster/r.viewshed.exposure/r.viewshed.exposure.py]: Addressed in c5c411a

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.py (hunk 9 above)
Comment (ninsbl) [reply src/raster/r.viewshed.exposure/r.viewshed.exposure.py:372]: It is indeed a workaround, see:
https://github.com/OSGeo/grass/issues/1436

Mentioned in comment in 294ce1d

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.py (hunk 10 above)
Comment (ninsbl) [reply src/raster/r.viewshed.exposure/r.viewshed.exposure.py:376]: Done in 294ce1d

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.py (hunk 13 above)
Comment (ninsbl) [reply src/raster/r.viewshed.exposure/r.viewshed.exposure.py:882]: Yes, thanks for the hin. This could indeed be solved with a reclass map ...

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.html [hunk 16]
```diff
@@ -0,0 +1,222 @@
+<h2>DESCRIPTION</h2>
```
[the remark this answers is not in the document]
Comment (ninsbl) [reply src/raster/r.viewshed.exposure/r.viewshed.exposure.html:1]: Addressed in eb32913

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.html [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+<i>Example of r.viewshed.exposure (2)</i>
+</div>
+
+<h2>TODO</h2>
+<ul>
+  <li>Implement variable exposure source height.</li>
+  <li>Implement possibility to switch between absolute and relative values
+    of visual exposure (now absolute).</li>
+</ul>
+<h2>KNOWN ISSUES</h2>
```
[the remark this answers is not in the document]
Comment (ninsbl) [reply src/raster/r.viewshed.exposure/r.viewshed.exposure.html]: Addressed in eb32913

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.py [hunk 18]
```diff
@@ -0,0 +1,1409 @@
...
+
+# %rules
+# % exclusive: sampling_points,source
+# %end
+
+# %option G_OPT_R_INPUT
+# % key: weights
+# % required: no
+# % label: Name of input raster map of viewshed weights
+# % guisection: Exposure source settings
```
[the remark this answers is not in the document]
Comment (ninsbl) [reply src/raster/r.viewshed.exposure/r.viewshed.exposure.py]: Implemented in ae50e34

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.py (hunk 11 above)
Comment (ninsbl) [reply src/raster/r.viewshed.exposure/r.viewshed.exposure.py]: Implemented in e45b649

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.py (hunk 12 above)
Comment (ninsbl) [reply src/raster/r.viewshed.exposure/r.viewshed.exposure.py]: I think it is a remnant from earlier progress tracking (right, @zofie-cimburova )... Removed it now in e45e875. We will see if it was needed...

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.py [hunk 19]
```diff
@@ -0,0 +1,1391 @@
...
+
+    # Sampling settings
+    source_sample_density = float(options["sample_density"])
+    seed = options["seed"]
+
+    if not seed:  # if seed is not set, set it to process number
+        seed = os.getpid()
+
+    # Optional
+    cores = int(options["cores"])
```
Comment (petrasovaa) [inline src/raster/r.viewshed.exposure/r.viewshed.exposure.py:1115]: The should be nprocs now, also needs to be changed in manual.

**Review by petrasovaa — CHANGES_REQUESTED** [review]

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (ninsbl) [reply src/raster/r.viewshed.exposure/r.viewshed.exposure.py]: For memory, the standard parser option is actually `G_OPT_MEMORYMB`.
(see: https://grass.osgeo.org/grass80/manuals/parser_standard_options.html)

`G_OPT_M_MEMORY` (or `G_OPT_M_MEMORYMB`) would be more consistent I would say, but there has probably been a reason for naming it that way?

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.py (hunk 19 above)
Comment (ninsbl) [reply src/raster/r.viewshed.exposure/r.viewshed.exposure.py]: Thanks. Fixed in 4b78c6e.

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.py (hunk 8 above)
Comment (ninsbl) [reply src/raster/r.viewshed.exposure/r.viewshed.exposure.py]: Implemented in 0e46fee

**Comment by ninsbl** [comment]
Note to self: remaining issues are:
1) Drop use of mask: https://github.com/OSGeo/grass-addons/pull/708#discussion_r819280761
2) Use kwargs: https://github.com/OSGeo/grass-addons/pull/708#discussion_r819264796

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.py (hunk 13 above)
Comment (ninsbl) [reply src/raster/r.viewshed.exposure/r.viewshed.exposure.py:880]: Looking at this again, the differences is the use of `r.random.cells` here, that does not have an input option (in contrast to `r.random` used in `r.sample.category`) So, i think we stick to using the mask for now, unless we rework the sampling more generally (i.e. using only raster and coordinates).

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.py [hunk 20]
```diff
@@ -0,0 +1,1409 @@
...
+            all_nan
+        ] = np.nan
+
+        counter += 1
+
+    return np_cum
+
+
+def binary(
+    lreg_shape, t_loc, np_viewshed, reg, exp_range, r_dsm, dsm_type, v_elevation, b_1
```
[the remark this answers is not in the document]
Comment (ninsbl) [reply src/raster/r.viewshed.exposure/r.viewshed.exposure.py:458]: This is a very good suggestion, and I should have thought about it myself earlier. Unfortunately, now, this requires more significant code changes. Since this affects mainly code quality (and less performance or results) I would be glad if we could merge the current state and add this (maybe together with more informative variable names in the numpy sections) as a todo for later (after the PhD defense).

Using keyword arguments (and maybe functools.partial) could simplify the code quite a bit, I guess. Note to self: Here are some possible approaches:
- kwargs with starmap: https://stackoverflow.com/questions/45718523/pass-kwargs-to-starmap-while-using-pool-in-python
- solution with functools.partial: https://github.com/ipython/ipyparallel/issues/147

The functools.partial approach seems to be the most appropriate to me (with regards to improvements in code quality). But we would have to try out one or the other...

**Comment by ninsbl** [comment]
Most requested changes have been addressed now. As mentioned I would be glad if we could address the kwargs issue (and the numpy variables) after merging a first version of the addon.

Would you mind having a second lokk at the current state? That would be great.

And thanks again for the thorough review. Very much appreciated!

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (petrasovaa) [reply src/raster/r.viewshed.exposure/r.viewshed.exposure.py]: I don't think there was any particular reason, they were added separately and probably nobody gave that much thought.

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.py (hunk 13 above)
Comment (petrasovaa) [reply src/raster/r.viewshed.exposure/r.viewshed.exposure.py:880]: Sure, I was not sure how different was this case from r.sample.category.

**Comment by petrasovaa** [comment]
> > What is little frustrating is how unreadable the numpy portions are (at least to me), but there is probably nothing to do about that.
> 
> Longer variable names may help, what do you think? Thought that will make the code longer...

It may help somewhat, perhaps also adding more comments. I will try to understand the actual algorithm, that will help as well.

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.py [hunk 21]
```diff
@@ -0,0 +1,1391 @@
...
+        # pygrass sets region for pygrass tasks
+        lreg = deepcopy(reg)
+        lreg.set_bbox(Bbox(loc_reg_n, loc_reg_s, loc_reg_e, loc_reg_w))
+        lreg.set_raster_region()
+
+        # Create processing environment with region information
+        c_env = os.environ.copy()
+        c_env["GRASS_REGION"] = grass.region_env(
+            n=loc_reg_n, s=loc_reg_s, e=loc_reg_e, w=loc_reg_w
+        )
```
Comment (petrasovaa) [inline src/raster/r.viewshed.exposure/r.viewshed.exposure.py:347]: When I run the test, I am getting:
```
WARNING: East-west resolution does not equal north-south resolution. The
         viewshed computation assumes the cells are square, so in this case
         this may result in innacuracies.
```

I am wondering if that's not caused by some incorrect alignment, shouldn't you use:
```suggestion
        c_env["GRASS_REGION"] = grass.region_env(
            n=loc_reg_n, s=loc_reg_s, e=loc_reg_e, w=loc_reg_w, align=r_dsm
        )
```

File: src/raster/r.viewshed.exposure/testsuite/test_r_viewshed_exposure.py [hunk 22]
```diff
@@ -0,0 +1,288 @@
...
+    }
+
+    @classmethod
+    def setUpClass(cls):
+        """Save the current region
+        We cannot use temp_region as it is used by the module.
+        """
+
+        # Save current region to temporary file
+        cls.runModule("g.region", flags="u", save="{}_region".format(cls.tempname))
```
Comment (petrasovaa) [inline src/raster/r.viewshed.exposure/testsuite/test_r_viewshed_exposure.py:165]: I don't see why use_temp_region wouldn't work here:

```suggestion
        gs.use_temp_region()
        gs.run_command("g.region", raster=cls.dsm)
```

Comment (ninsbl) [reply src/raster/r.viewshed.exposure/testsuite/test_r_viewshed_exposure.py]: We tried that earlier but it failed with raster2numpy or numpy2raster (don`t remember)...
But lets try again with G8 in 43116a4 ...

File: src/raster/r.viewshed.exposure/r.viewshed.exposure.py (hunk 21 above)
Comment (zofie-cimburova) [reply src/raster/r.viewshed.exposure/r.viewshed.exposure.py:361]: > When I run the test, I am getting:
> 
> ```
> WARNING: East-west resolution does not equal north-south resolution. The
>          viewshed computation assumes the cells are square, so in this case
>          this may result in innacuracies.
> ```
> 
> I am wondering if that's not caused by some incorrect alignment, shouldn't you use:

The alignment didn't work properly if sampling points were defined manually (i.e. if they were not exactly in the middle of a raster cell). Corrected in https://github.com/OSGeo/grass-addons/pull/708/commits/4ae225ed361d80c8ae7e4ccc0d5c8d2bff46e726

**Review by petrasovaa — APPROVED** [review]
Thanks for addressing all the comments!
